src/Host/Broker/Impl/RemoteUri/RemoteUriHelper.cs

```diff
@@ -5,37 +5,88 @@
 using System.Collections.Generic;
 using System.IO;
 using System.Net;
+using System.Net.WebSockets;
+using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Http;
 using Microsoft.Common.Core;
 using Microsoft.R.Host.Protocol;
 
 namespace Microsoft.R.Host.Broker.RemoteUri {
     public class RemoteUriHelper {
+
+        private static async Task DoWebSocketReceiveSendAsync(WebSocket receiver, WebSocket sender, CancellationToken ct) {
+            if (receiver == null || receiver == null) {
+                return;
+            }
+
+            ArraySegment<byte> receiveBuffer = WebSocket.CreateServerBuffer(65335);
+            while (receiver.State == WebSocketState.Open && sender.State == WebSocketState.Open) {
+                if (ct.IsCancellationRequested) {
+                    Task.WhenAll(
+                        receiver.CloseAsync(WebSocketCloseStatus.NormalClosure, "Normal closure", CancellationToken.None),
+                        sender.CloseAsync(WebSocketCloseStatus.NormalClosure, "Normal closure", CancellationToken.None))
+                        .DoNotWait();
+                    return;
+                }
+
+                WebSocketReceiveResult result = await receiver.ReceiveAsync(receiveBuffer, ct);
+
+                byte[] data = null;
+                using (MemoryStream ms = new MemoryStream()) {
+                    await ms.WriteAsync(receiveBuffer.Array, receiveBuffer.Offset, result.Count);
+                    await ms.FlushAsync();
+                    data = ms.ToArray();
+                }
+
+                ArraySegment<byte> sendBuffer = new ArraySegment<byte>(data);
+                await sender.SendAsync(sendBuffer, result.MessageType, result.EndOfMessage, ct);
+
+                if (result.MessageType == WebSocketMessageType.Close) {
+                    await receiver.CloseAsync(WebSocketCloseStatus.NormalClosure, "Normal closure", ct);
```

Comment:
I believe this is supposed to be an empty string normally - message is for errors and such.